kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -277,144 +269,124 @@ func (c *loadBalancerController) handleIPPoolUpdate(kvp model.KVPair) {
+									log.Errorf("Error releasing IP(%s) for svc: %s", ingressIP.IP, svc.Name)
+								}
 							}
 						}
 					}
-				}
-			} else {
-				// Service no longer exists, leak confirmed. Release all IPs allocated with the specific handle
-				log.Infof("Service not found for handle: %s. Releasing unused IPs", *attr.AttrPrimary)
-				err := c.releaseIP(*attr.AttrPrimary)
-				if err != nil {
-					log.Errorf("Error releasing IPAM for handle %s: %s", *attr.AttrPrimary, err)
+				} else {
+					// Service no longer exists, leak confirmed. Release all IPs allocated with the specific handle
+					log.Infof("Service not found for handle: %s. Releasing unused IPs", *attr.AttrPrimary)
+					err := c.releaseIP(*attr.AttrPrimary)
+					if err != nil {
+						log.Errorf("Error releasing IPAM for handle %s: %s", *attr.AttrPrimary, err)
+					}
 				}
 			}
 		}
 	}
 
 	// Check that all services have assigned IPs as requested, skip if there are no ippools
 	if len(c.ipPools) != 0 {
-		for _, key := range c.lbResourceCache.ListKeys() {
-			obj, empty := c.lbResourceCache.Get(key)
-			if !empty {
-				service := obj.(*v1.Service)
-				if isCalicoManagedLoadBalancer(service, c.cfg.AssignIPs) {
-					if service.Status.LoadBalancer.Ingress == nil ||
-						(len(service.Status.LoadBalancer.Ingress) == 1 &&
-							(*service.Spec.IPFamilyPolicy == v1.IPFamilyPolicyRequireDualStack) || *service.Spec.IPFamilyPolicy == v1.IPFamilyPolicyPreferDualStack) {
-						err := c.assignIP(service)
-						if err != nil {
-							log.Errorf("Error assigning IP to service %s: %s", service.Name, err)
-						}
-					}
+		for _, svc := range services {
+			if svc.Status.LoadBalancer.Ingress == nil ||
+				(len(svc.Status.LoadBalancer.Ingress) == 1 &&
+					(*svc.Spec.IPFamilyPolicy == v1.IPFamilyPolicyRequireDualStack) || *svc.Spec.IPFamilyPolicy == v1.IPFamilyPolicyPreferDualStack) {
+				err = c.assignIP(&svc)
+				if err != nil {
+					log.WithError(err).Errorf("Error assigning IP to svc: %s", svc.Name)
 				}
 			}
 		}
 	}
-}
-
-func (c *loadBalancerController) runWorker() {
-	for c.processNextItem() {
-	}
-}
-
-func (c *loadBalancerController) processNextItem() bool {
-	workqueue := c.lbResourceCache.GetQueue()
-	key, quit := workqueue.Get()
-	if quit {
-		return false
-	}
-
-	err := c.syncToDatastore(key.(string))
-	c.handleErr(err, key.(string))
 
-	workqueue.Done(key)
-	return true
-}
-
-// syncToDatastore processes the next item in queue
-func (c *loadBalancerController) syncToDatastore(key string) error {
-	obj, exists := c.lbResourceCache.Get(key)
-
-	if !exists {
-		// Service does not exist release the IPs
-		return c.releaseIP(key)
-	} else {
-		svc := obj.(*v1.Service)
-		// We don't have IP, assign one
-		if svc.Status.LoadBalancer.Ingress == nil {
-			return c.assignIP(svc)
-		} else {
-			// Service was updated
-			return c.checkStatus(svc)
+	if len(c.servicesToUpdate) != 0 {
+		for key, svcObj := range c.servicesToUpdate {
+			switch svcObj.updateType {
```

Comment:
This is somewhat hastily written out, so may not be bulletproof, but just as something to run past you to see what you think - the following pseudocode is probably where I would start from if trying to figure out this logic myself. 

---

The controller would have some internal state that it uses to efficiently track IP allocation state. It would probably need to initialize these maps on startup with IPAM state, and then make incremental changes to them as it allocates and releases IP addresses.

Could use a helper structure to encapsulate that logic, which might make it a bit tidier:

```go
type allocationTracker struct {
  // Keep track of IP allocations by Service, as well as a 
  // reverse lookup of service by IP. These maps are populated based on the IPAM DB state 
  // and NOT by the Service objects.
  servicesByIP map[string]string
  ipsByService map[string]map[string]bool
}

func (t *allocationTracker) AssignAddress(svc *v1.Service, ip string)
func (t *allocationTracker) ReleaseAddress(svc *v1.Service, ip string)
func (t *allocationTracker) DeleteService(svc *v1.Service)
```

The `syncIPAM` function would have something like this:

```go
// Go through each Service that we know about and make sure it is properly in sync.
for _, svc := range c.informer.List() {
  if err := c.syncService(svc); err != nil {
    logrus.WithError(err).Warn("warning")
  }
}
```

```go
// syncService does the following:
// - Releases any IP addresses in the IPAM DB associated with the Service that are not in the Service status.
// - Allocates any addresses necessary to satisfy the Service LB request
// - Updates the controllers internal state tracking of which IP addresses are allocated.
// - Updates the IP addresses in the Service Status to match the IPAM DB.
func syncService(svc *v1.Service) error {
  // TODO
  return nil
}
```

you could even have "event driven" logic accept incoming service updates funneled into the same function, simply calling `syncService` when we get service updates:

```go
select {
  case <-kickChan:
    syncIPAM();
  case <- syncerUpdates:
    handleUpdate(upd)
  case svcID <- serviceUpdates:
    svc := c.informer.Get(svcID)
    syncService(svc)
}
```
}